src/Arcus.WebApi.Security/Authentication/SharedAccessKeyAuthenticationAttribute.cs

```diff
@@ -0,0 +1,104 @@
+    {
+        private readonly string _headerName, _secretName;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SharedAccessKeyAuthenticationAttribute"/> class.
+        /// </summary>
+        /// <param name="headerName">The name of the request header which value must match the stored secret.</param>
+        /// <param name="secretName">The name of the secret that's being retrieved using the <see cref="ISecretProvider.Get"/> call.</param>
+        public SharedAccessKeyAuthenticationAttribute(string headerName, string secretName)
+        {
+            if (String.IsNullOrWhiteSpace(headerName))
+            {
+                throw new ArgumentException("Header name cannot be blank", nameof(headerName));
+            }
+
+            if (String.IsNullOrWhiteSpace(secretName))
+            {
+                throw new ArgumentException("Secret name cannot be blank", nameof(secretName));
+            }
+
+            _headerName = headerName;
+            _secretName = secretName;
+        }
+        
+        /// <inheritdoc />
+        public override async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.HttpContext?.Request == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request", 
+                    nameof(context.HttpContext.Request));
+            }
+
+            if (context.HttpContext.Request.Headers == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request headers",
+                    nameof(context.HttpContext.Request.Headers));
+            }
+
+            if (context.HttpContext.RequestServices == null)
+            {
+                throw new ArgumentException(
+                    "Invalid action context given without any HTTP request services", 
+                    nameof(context.HttpContext.RequestServices));
+            }
+
+            var userDefinedSecretProvider = context.HttpContext.RequestServices.GetService<ISecretProvider>();
+            if (userDefinedSecretProvider == null)
+            {
+                throw new KeyNotFoundException(
+                    $"No configured {nameof(ISecretProvider)} implementation found in the request service container. "
+                    + "Please configure such an implementation (ex. in the Startup) of your application");
+            }
+
+            var cachedSecretProvider = new CachedSecretProvider(userDefinedSecretProvider);
+            string foundSecret = await cachedSecretProvider.Get(_secretName);
+            if (foundSecret == null)
+            {
+                throw new AuthenticationException(
```

Comment:
Do we have a SecretNotFoundException here or in Arcus Security?

Reply:
Yes, we have! I'll use that.

Reply:
Great, thanks! :+1: